resharper/src/resharper-unity/CSharp/Daemon/Stages/PerformanceCriticalCodeAnalysis/PerformanceCriticalCodeAnalysisStage.cs

```diff
@@ -0,0 +1,560 @@
+                #endregion
+            }
+
+            public bool InteriorShouldBeProcessed(ITreeNode element, HotMethodAnalyzerContext context)
+            {
+                return !(element is ICSharpClosure);
+            }
+
+            public bool IsProcessingFinished(HotMethodAnalyzerContext context)
+            {
+                return false;
+            }
+
+            public void ProcessBeforeInterior(ITreeNode element, HotMethodAnalyzerContext context)
+            {
+                // handle depth of analysis
+                if (element is IInvocationExpression)
+                    context.Depth++;
+                
+                if (element is ICSharpTreeNode node)
+                    node.Accept(this, context);
+            }
+
+            public void ProcessAfterInterior(ITreeNode element, HotMethodAnalyzerContext context)
+            {
+                // handle depth of analysis
+                if (element is IInvocationExpression)
+                    context.Depth--;
+            }
+            
+            #region data
+            
+            private static readonly ISet<string> ourKnownComponentCostlyMethods = new HashSet<string>()
+            {
+                "GetComponentInChildren",
+                "GetComponentInParent",
+                "GetComponentsInChildren",
+                "GetComponent",
+                "GetComponents",
+            };
+            
+            private static readonly ISet<string> ourKnownGameObjectCostlyMethods = new HashSet<string>()
+            {
+                "Find",
+                "FindGameObjectsWithTag",
+                "FindGameObjectWithTag",
+                "FindWithTag",
+                "GetComponent",
+                "GetComponents",
+                "GetComponentInChildren",
+                "GetComponentInParent",
+                "GetComponentsInChildren",
+            };
+            
+            private static readonly ISet<string> ourKnownTransformCostlyMethods = new HashSet<string>()
+            {
+                "Find"
+            };
+            
+            private static readonly ISet<string> ourKnownResourcesCostlyMethods = new HashSet<string>()
+            {
+                "FindObjectsOfTypeAll",
+            };
+            
+            private static readonly ISet<string> ourKnownObjectCostlyMethods = new HashSet<string>()
+            {
+                "FindObjectsOfType",
+                "FindObjectOfType",
+                "FindObjectsOfTypeIncludingAssets",
+            };
+            
+            #endregion
+        }
+
+        private class HotMethodAnalyzerContext
+        {
+            // Current depth of analysis
+            public int Depth { get; set; }
+            
+            // Inverted call graph for costly reachable mark propagation after visit AST
+            public IReadOnlyDictionary<IDeclaredElement, HashSet<IDeclaredElement>> InvertedCallGraph 
```

Comment:
Creating a copy of the whole dictionary isn't very good in terms of memory consumption. Maybe return the original dictionary?